pkg/proc/evalop/evalcompile.go

```diff
@@ -0,0 +1,553 @@
+package evalop
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/constant"
+	"go/printer"
+	"go/token"
+	"strconv"
+	"strings"
+
+	"github.com/go-delve/delve/pkg/dwarf/godwarf"
+	"github.com/go-delve/delve/pkg/dwarf/reader"
+)
+
+var (
+	ErrFuncCallNotAllowed = errors.New("function calls not allowed without using 'call'")
+)
+
+type compileCtx struct {
+	evalLookup
+	ops        []Op
+	allowCalls bool
+	curCall    int
+}
+
+type evalLookup interface {
+	FindTypeExpr(ast.Expr) (godwarf.Type, error)
+	HasLocal(string) bool
+	HasGlobal(string, string) bool
+	HasBuiltin(string) bool
+	LookupRegisterName(string) (int, bool)
+}
+
+func Compile(lookup evalLookup, t ast.Expr) ([]Op, error) {
+	ctx := &compileCtx{evalLookup: lookup, allowCalls: true}
+	err := ctx.compileAST(t)
+	if err != nil {
+		return nil, err
+	}
+
+	err = ctx.depthCheck(1)
+	if err != nil {
+		return ctx.ops, err
+	}
+	return ctx.ops, nil
+}
+
+func CompileSet(lookup evalLookup, lhe, rhe ast.Expr) ([]Op, error) {
+	ctx := &compileCtx{evalLookup: lookup, allowCalls: true}
+	err := ctx.compileAST(rhe)
+	if err != nil {
+		return nil, err
+	}
+
+	if isStringLiteral(rhe) {
+		err = ctx.compileAllocLiteralString()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = ctx.compileAST(lhe)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.addOp(&SetValue{lhe: lhe, Rhe: rhe})
+
+	err = ctx.depthCheck(0)
+	if err != nil {
+		return ctx.ops, err
+	}
+	return ctx.ops, nil
+
+}
+
+func (ctx *compileCtx) compileAllocLiteralString() error {
+	ctx.addOp(&CallInjectionAllocString{Phase: 0})
+	ctx.addOp(&CallInjectionAllocString{Phase: 1})
+	ctx.addOp(&CallInjectionAllocString{Phase: 2})
+	return nil
+}
+
+func (ctx *compileCtx) addOp(op Op) {
+	ctx.ops = append(ctx.ops, op)
+}
+
+func (ctx *compileCtx) depthCheck(endDepth int) error {
```

Comment:
I know this is an internal consistency check but can you add some more documentation around it so it's clearer further down the line when reading the code in the future.